stravalib/client.py

```diff
@@ -1906,16 +1673,16 @@ def _fill_buffer(self):
         # If we cannot fetch anymore from the server then we're done here.
         if self._all_results_fetched:
             self._eof()
-
         raw_results = self.result_fetcher(
             page=self._page, per_page=self.per_page
         )
 
         entities = []
         for raw in raw_results:
             try:
-                new_entity = self.entity.parse_obj(raw)
-                # TODO: try adding bind_client to entity
+                new_entity = self.entity.parse_obj(
+                    {**raw, **{'bound_client': self.bind_client}}
+                )
```

Comment:
Just to be sure, binds the client to any entity fetched in `BatchedResultIterator`

Reply:
please tell me if i understand this correctly. When you initially instantiate a Client object, it is empty and a workflow slowly adds content to Client attributes such as authentication. Once you download activities you get a BatchedResult Iterator. So when you say bind, does this just mean that the activity object (the iterator) is then "connected to" or a part of the Client object? i'm a bit confused because the `client.get_activities()` returns an iterator but it's not necessarily a part of the `client` object created. just trying to understand language if you don't mind!

Reply:
> i'm a bit confused because the client.get_activities() returns an iterator but it's not necessarily a part of the client object created

It's the other way around: the `client` object becomes a part of the entities (or iterators of entities) it returns.

In this case, when a `BatchedResultsIterator` is created (e.g., as part of `client.get_activities()`), the client attaches itself to this iterator as follows:

```python
return BatchedResultsIterator(
            entity=model.Activity,
            bind_client=self,  # <-- client "binds" itself to the iterator
            result_fetcher=result_fetcher,
            limit=limit,
        )
```

This is not hitting the API yet. Now, when we request the next batch of results from this iterator, e.g., by `next()` or `list()`, it internally will hit the Strava API to fetch entities (see `_fill_buffer(self)`) and attach its client reference (pointed to in the snippet above) to each of these entities.

By making a `client` object part of an `Activity` object, we can now use constructs like these:

```python
@lazy_property
    def comments(self):
        return (
            self
            .bound_client  # <-- this is the same "bind_client=self" object from the snippet above
            .get_activity_comments(self.id)
        )
```

which makes it appear to the outside world that `comments` is a plain attribute of an `Activity` object, while in fact, referring to `some_activity.comments` will hit the API through the bound client.

Reply:
ok got it. thank you for taking the time to help me understand the way these objects interact!!